### ISSUE TITLE
Pushing with an application template without a deployment stalls

### ISSUE THREAD

**Issue by agracey** [epinio/epinio#1629]
**Describe the bug**

If I edit the deployment manifest to use a Statefulset instead of a Deployment, the push command doesn't exit after the build and deploy. I can tell from kubectl that the application is running correctly (and can access the app) so the only issue is that the CLI doesn't know that the app is alive.

**To Reproduce**
Add a new application template with my statefulset template: https://chameleons.sfo3.digitaloceanspaces.com/epinio-stateful-application-0.1.15.tgz 

Push an application with this template.

**Expected behavior**
I would expect the CLI to return successfully when the application is running

**Comment by andreas-kupries** [comment]
IIRC we have places in the CLI which expect a `Deployment` resource to query and/or modify.
I.e. they are :warning: __hardwired__ to query/change a `Deployment` resource for the app.

The `check if the app is running` action is one of them. It IIRC queries for existence of the `Deployment` first, and then the status field until it is `ready`.

Your push is actually not stalled. It is waiting patiently for the `Deployment` to exist, until timeout.
That timeout should be 3 minutes in v1. It was 10 minutes until shortly before the release of v1.

This ticket is essentially a request to modify Epinio to divorce itself from the notion that the central resource of a running app is always a `Deployment` ?

The obvious question here is, how do we tell Epinio what kind of central resource is used by a custom chart ?
And then, what kinds are there we should/could support ? (I.e. beyond `Deployment` and `StatefulSet`).

**Comment by andreas-kupries** [comment]
The cli side check is https://github.com/epinio/epinio/blob/4d05cce028b5899e720502a74268bf1814447fcb/internal/cli/usercmd/push.go#L237-L243

It reaches server side https://github.com/epinio/epinio/blob/4d05cce028b5899e720502a74268bf1814447fcb/internal/api/v1/application/running.go#L42-L45

Except that it should semi fail earlier, in `application.Lookup`, where it should fail to fetch the Deployment resource: 
https://github.com/epinio/epinio/blob/4d05cce028b5899e720502a74268bf1814447fcb/internal/application/workload.go#L110-L112

And the caller pulls important information out of the resource, see https://github.com/epinio/epinio/blob/4d05cce028b5899e720502a74268bf1814447fcb/internal/application/workload.go#L197-L205

And confirmed: The server log contains

```
2022/07/21 11:55:50 EpinioServer: "msg"="ERROR" "error"="no status available for application without workload" "requestId"="bb6dcdb6-f33d-4446-986c-93ada6cff033" "origin"="/api/v1/namespaces/workspace/applications/fox/running" "error"="no status available for application without workload"
```

Further conformed that removing the cli check does get us over the problem reported here.

However every API endpoint operating on/with the app workloads will break in various ways.
I expect that most will complain that the app is not running, or simply not provide data.

**Comment by andreas-kupries** [comment]
PR created. ... Now going into spike mode. Investigate how to get rid of the other places locked to `Deployment` as main app resource.

**Comment by andreas-kupries** [comment]
@agracey  Note: Your chart is based on 0.1.15 ... That is from before the rename of `epinio.suse.org` to `epinio.io`. This will break other parts of the system also.

**Comment by andreas-kupries** [comment]
Investigating the CI failures ... It looks I got the new determination of the `workload name` wrong.

In the main branch the workload name is the name of the Deployment responsible for the Pod. This is also the name of the app's main `Container` in the `Pod`. Tests use that for `kubectl exec`. Might also be used for `epinio exec`.

In my change I used the owner reference, and that points to the `ReplicaSet` sitting between `Deployment` and `Pod` instead.

The pod has the proper name only in the container-related fields, which will not help for Pods with multiple containers.
Not sure if I can recurse up the ownership links ... Would have to handle all the different kinds of resources above Pod and the Deployment equivalent. Further, what would the stop condition be ?

Trying to infer it from the pod name, by repeatedly stripping a trailing `-.*` segment would sort of equivalent to the recursion, purely syntactically. Stop condition would be finding a container with the name ... Hackish. Likely brittle.

That said, the name is generated in the app chart, and then set into the various places (deployment name, and `spec.template.container[...].name`) ... Set it as a label FOO in the pod as well, and the make presence and semantics of FOO a requirement for any working app chart.

While that imposes a condition on user-written app charts, that condition should not be too onerous.
Trialing this now.

**Comment by andreas-kupries** [comment]
Continued CI investigation ... The various `app env` test failing ... It tries to restart an app which is not running. It misidentifies it as having a workload ... I am mishandling the case of `no pods present`. This is the new condition for `no workload`.

Locally the failed tests seem to all work now.
Pushed all changes.

**Comment by agracey** [comment]
There needs to be agreement between what we expect and what an operator writes so this restriction makes sense. We just need to document this in https://docs.epinio.io/howtos/create_custom_appcharts so there's no surprises 

We could improve the usability but maybe providing an example with something like knative would make it easier to see how to pull everything together.

**Comment by andreas-kupries** [comment]
New documentation issue: https://github.com/epinio/docs/issues/151

**Comment by jimmykarily** [comment]
@agracey is it possible to have a chart that uses Statefulsets implemented (maybe as part of this: https://github.com/epinio/epinio/issues/1273) so that it can be used in an automated test for this fix/change?

**Comment by agracey** [comment]
Yes, I have one already built so I'll push it as soon as I remember where the code is 😆
